sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala

```diff
@@ -85,6 +85,12 @@ private[sql] class JSONOptions(
 
   val multiLine = parameters.get("multiLine").map(_.toBoolean).getOrElse(false)
 
+  /**
+   * Standard charset name. For example UTF-8, UTF-16 and UTF-32.
+   * If charset is not specified (None), it will be detected automatically.
```

Comment:
Does this detect the encoding for newlines too?

Reply:
Do you mean the encoding of records/lines delimiter? It depends on the mode. In multiline mode, jackson is able to do that. In the case of per-line mode, Hadoop LinerRecordReader could accept delimiters in any charsets but by defaults it splits input by `'\r'`, `'\n'`, and `'\r\n'` in UTF-8. This will be fixed in separate PRs for the issues: https://issues.apache.org/jira/browse/SPARK-23724 and https://issues.apache.org/jira/browse/SPARK-23725

Reply:
Shall we fix that first with text datasource since schema inference in JSON is dependent on Text datasource? You are exposing incomplete option now.

Reply:
> schema inference in JSON is dependent on Text datasource

Could you clarify this, please. It is not completely clear to me what do you mean.

Reply:
See this https://github.com/apache/spark/commit/8fb2a02e2ce6832e3d9338a7d0148dfac9fa24c2. It uses Text datasource when it loads lines when we infer schema.

If we fix encodings with newline first, it's required to Text datasource first I believe.

Reply:
ok. How this one helps to solve the problem that I am trying to solve by this PR: jackson's charset auto-detection mechanism can fail on even UTF-8 encoding and can infer wrong charset (see https://github.com/apache/spark/pull/20302) due to many reasons. And an user doesn't have any possibilities to fix the issue and bypass the auto-detection.

Reply:
Json's schema inference use the text datasource to separate the lines before we go through jackson parser where the charset for newlines should be respected. Shouldn't we better fix text datasource with the hadoop's line reader first?

Reply:
A fix in hadoop line reader and this PR solve 2 different problem. Any fix in hadoop line reader will not fix the problem of wrong encoding detection. I don't understand why this PR must depend on a fix in line reader. I would say a custom record separator will solve newline problem too (https://issues.apache.org/jira/browse/SPARK-23724). 

> Shouldn't we better fix text datasource with the hadoop's line reader first?

Could you tell me how this PR blocks solving the problem in Hadoop's LineReader?

Reply:
> Could you tell me how this PR blocks solving the problem in Hadoop's LineReader?

Because the exposed `charset` option is incomplete here because the encodings.
Also, I want to see how we can solve that problem in SPARK-23724 first too. I am actually not quite worried of the whole changes proposed here for now.

Why don't we just fix that problem first if you plan to fix both eventually anyway?

Reply:
I would like to introduce changes step-by-step to eliminate creating of a "patch bomb". Regarding to SPARK-23724, I am going to propose this PR: https://github.com/MaxGekk/spark-1/pull/1 /cc @hvanhovell

Reply:
Yea, step-by-step. I am not suggesting to fix all here. I meant SPARK-23724 should be fixed first with text data source.